shared/modules/factory.go

```diff
@@ -4,25 +4,31 @@ package modules
 // a variadic `ModuleOption` argument(s) and returns a `Module`and an error.
 type ModuleFactoryWithOptions FactoryWithOptions[Module, ModuleOption]
```

Comment:
Should we add other types like `type ModuleFactory Factory[Module]` as well?

Reply:
:thinking: I could imagine an argument for either case but without being exhaustive here, I think it's nice that all modules conform to the `ModuleFactoryWithOptions` interface so that it can be embedded in `Module` (and therefore consistent between modules). Additionally, for any module which has a required config, it should load that from the base config file (unless I'm mistaken, I believe every module currently does this). This eliminates the possibility of needing a `ModuleFactoryWithConfig` or `ModuleFactoryWithConfigAndOptions`, and since the variadic arg on `ModuleFactoryWithOptions` can be omitted, the DevEx would be equivalent on the consumer side as compared to using a `ModuleFactory` interface like you suggest. The only difference then would be on the module side, we're basically requiring that each module accept and optionally consider a slice of `ModuleOption`s. I think it's also more prudent to be more forward compatible at the module-level (as compared with submodules), which makes me prefer this consistency accross modules.

Reply:
To add one more bit of detail; unlike `Module`s, requiring all submodules to accept and optionally consider a slice of options adds more boilerplace in the context of a submodule because each submodule will have to define its own option type for use with the `Factory...` generic type params, which it then may ignore.

Reply:
> > . Additionally, for any module which has a required config, it should load that from the base config file (unless I'm mistaken, I believe every module currently does this).
> 
> It "should", but there is no enforcement, correct? Not a request to change it, just confirming.

Correct. A module may or may not use any part of the base config via the runtime manager, via the bus but there is no enforcement that any particular module consults any particular part of the config (or doesn't).

> > I think it's also prudent to be more forward compatible at the module-level (as compared with submodules), which makes me prefer this consistency accross modules.
> 
> In #810, can you talk more about the need to differentiate modules and submodules? The only difference I currently see is where the interface is defined (shared or non-shared) so am thinking that an internal type cast might be able to consolidate the business logic.

:thinking: The main hangup I have is forcing submodules to conform to the `ModuleFactoryWithOptions`, this is what causes the need for the additional boilerplate around submodule dependency injection. Perhaps this could be addressed simply by removing the `Create()` method from `InitializableModule`  and defining a new `Submodule` interface type which doesn't embed `ModuleFactoryWithOptions` (nor `InterruptableModule`):

```go
type InitializableModule interface {
	// GetModuleName returns the name of the module.
	GetModuleName() string
}

type Module interface {
	InitializableModule
	IntegratableModule
	InterruptableModule
	ModuleFactoryWithOptions
}

type Submodule interface {
  InitializableModule
  IntegratableModule
}
```

Then `Submodule` can be enforced on the submodule alongside its local "factory" type. Wdyt?

Reply:
I updated #810 to include the `Submodule` interface now that you got me to think about it a bit more. I also updated/added some deliverables.

Reply:
Closing out this comment in favor to continue the conversation in #810